pkg/core/diamond_commit.go

```diff
@@ -0,0 +1,326 @@
+package core
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	iradix "github.com/hashicorp/go-immutable-radix"
+	"github.com/oneconcern/datamon/pkg/core/status"
+	"github.com/oneconcern/datamon/pkg/errors"
+	"github.com/oneconcern/datamon/pkg/model"
+	"github.com/oneconcern/datamon/pkg/storage"
+	"github.com/segmentio/ksuid"
+	"go.uber.org/zap"
+)
+
+// Commit a diamond
+func (d *Diamond) Commit(opts ...Option) error {
+	return d.implCommit(opts...)
+}
+
+// Cancel a diamond
+func (d *Diamond) Cancel() error {
+	// TODO(fred)
+	return status.ErrNotImplemented.WrapMessage("diamond cancel")
+}
+
+// WithState returns the diamond with its state modified and transition timestamp updated
+func (d *Diamond) WithState(state model.DiamondState) *Diamond {
+	d.DiamondDescriptor.State = state
+	d.DiamondDescriptor.StateChangeTime = time.Now()
+
+	switch state {
+	case model.DiamondDone, model.DiamondCancelled:
+		d.DiamondDescriptor.EndTime = d.DiamondDescriptor.StateChangeTime
+	}
+	return d
+}
+
+// collectSplits which are done, sorted by completion time
+func (d *Diamond) collectSplits(opts ...Option) (model.SplitDescriptors, error) {
+	splits := make(model.SplitDescriptors, 0, typicalSplitsNum)
+
+	err := ListSplitsApply(
+		d.RepoID,
+		d.DiamondDescriptor.DiamondID,
+		d.contextStores,
+		func(sd model.SplitDescriptor) error {
+			if sd.State == model.SplitDone {
+				splits = append(splits, sd)
+			}
+			return nil
+		},
+		opts...,
+	)
+	if err != nil {
+		return nil, errors.New("couldn't collect splits").Wrap(err)
+	}
+	return splits, nil
+}
+
+func (d *Diamond) checkBundleID() error {
+	if d.BundleID == "" {
+		return nil
+	}
+
+	// case of bundleID preservation (e.g. data migration use case)
+	id, err := ksuid.Parse(d.BundleID)
+	if err != nil {
+		return status.ErrInvalidKsuid.Wrap(err)
+	}
+	d.setBundleID(id.String())
+	exists, err := d.Exists(d.contexter())
+	if err != nil {
+		return errors.New("could not check for bundle existence").Wrap(err)
+	}
+	if exists {
+		return status.ErrBundleIDExists.WrapMessage("bundleID: %v", id)
+	}
+	return nil
+}
+
+func (d *Diamond) implCommit(opts ...Option) (err error) {
+	// check if repo exists
+	if err = RepoExists(d.RepoID, d.contextStores); err != nil {
+		return err
+	}
+
+	// check when bundle is already specified (e.g. data migration use case)
+	if err = d.checkBundleID(); err != nil {
+		return err
+	}
+
+	logger := d.l.With(zap.String("diamond_id", d.DiamondDescriptor.DiamondID))
+
+	// critical section #1
+	// TODO(fred): better protect critical section
```

Comment:
recording consideration of state-management metaphors for GCS while here for future reference:

do lockfiles translate to gcs?

"mutations" on immutable objects are implemented by creating a new object, and immutability has some desirable properties.  so state could be tracked via a directory-like prefix `.../<diamond-prefix>/<state-name>/` and then use some lexicographical ordering (e.g. zero-padded integers) on objects to distinguish the current state from previous states.

Reply:
yes, using such resources as "lock-files" was my initial design.
However, since we can't delete the files, and since state is not always going forward but may revert (from running to initialized, from failed to running, etc), I gave up that way.
The downside of this is now I have a race condition on updating gcs files.